Source/WarpX.cpp

```diff
@@ -2893,57 +2893,23 @@ void
 WarpX::BuildBufferMasksInBox ( const amrex::Box tbx, amrex::IArrayBox &buffer_mask,
                                const amrex::IArrayBox &guard_mask, const int ng )
 {
-    bool setnull;
-    const amrex::Dim3 lo = amrex::lbound( tbx );
-    const amrex::Dim3 hi = amrex::ubound( tbx );
-    Array4<int> msk = buffer_mask.array();
-    Array4<int const> gmsk = guard_mask.array();
-#if defined(WARPX_DIM_1D_Z)
-    int k = lo.z;
-    int j = lo.y;
-    for (int i = lo.x; i <= hi.x; ++i) {
-        setnull = false;
-        // If gmsk=0 for any neighbor within ng cells, current cell is in the buffer region
-        for (int ii = i-ng; ii <= i+ng; ++ii) {
-            if ( gmsk(ii,j,k) == 0 ) setnull = true;
-        }
-        if ( setnull ) msk(i,j,k) = 0;
-        else           msk(i,j,k) = 1;
-    }
-#elif defined(WARPX_DIM_XZ) || defined(WARPX_DIM_RZ)
-    int k = lo.z;
-    for     (int j = lo.y; j <= hi.y; ++j) {
-        for (int i = lo.x; i <= hi.x; ++i) {
-            setnull = false;
-            // If gmsk=0 for any neighbor within ng cells, current cell is in the buffer region
-            for     (int jj = j-ng; jj <= j+ng; ++jj) {
-                for (int ii = i-ng; ii <= i+ng; ++ii) {
-                    if ( gmsk(ii,jj,k) == 0 ) setnull = true;
-                }
-            }
-            if ( setnull ) msk(i,j,k) = 0;
-            else           msk(i,j,k) = 1;
-        }
-    }
-#elif defined(WARPX_DIM_3D)
-    for         (int k = lo.z; k <= hi.z; ++k) {
-        for     (int j = lo.y; j <= hi.y; ++j) {
-            for (int i = lo.x; i <= hi.x; ++i) {
-                setnull = false;
-                // If gmsk=0 for any neighbor within ng cells, current cell is in the buffer region
-                for         (int kk = k-ng; kk <= k+ng; ++kk) {
-                    for     (int jj = j-ng; jj <= j+ng; ++jj) {
-                        for (int ii = i-ng; ii <= i+ng; ++ii) {
-                            if ( gmsk(ii,jj,kk) == 0 ) setnull = true;
-                        }
+    auto const& msk = buffer_mask.array();
+    auto const& gmsk = guard_mask.const_array();
+    amrex::Dim3 ng3 = amrex::IntVect(ng).dim3();
+    amrex::ParallelFor(tbx, [=] AMREX_GPU_DEVICE (int i, int j, int k)
+    {
+        for         (int kk = k-ng3.z; kk <= k+ng3.z; ++kk) {
+            for     (int jj = j-ng3.y; jj <= j+ng3.y; ++jj) {
+                for (int ii = i-ng3.x; ii <= i+ng3.x; ++ii) {
+                    if ( gmsk(ii,jj,kk) == 0 ) {
+                        msk(i,j,k) = 0;
+                        return;
```

Comment:
Is it correct to have the `return` statement in the innermost loop over `ii` here? I'm wondering what happens if `gmsk(ii,jj,kk) == 0` is true for more than one combination of indices `(ii,jj,kk)`. Once the first combination is encountered, the corresponding `msk(i,j,k)` is set to zero and then the function "returns". Does it then reach the second combination of indices `(ii,jj,kk)` such that `gmsk(ii,jj,kk)=0`? Maybe I don't understand exactly where `return` brings us in this case.

Reply:
Looks like a bug to me, too.
```suggestion
```

Reply:
Once it reaches `return`, the iteration will stop and the rest of gmsk values will not have any impact. It's equivalent to
```
if (gmsk(i-1,j-1,k-1) == 0) {
    msk(i,j,k) = 0;
} else if (gmsk(i,j-1,k-1) == 0) {
    msk(i,j,k) = 0;
} else if (...) {
    msk(i,j,k) = 0;
}
....
else if (gmsk(i+1,j+1,k+1) == 0) {
    msk(i,j,k) = 0;
else {
    msk(i,j,k) = 1;
}
```
I think this is equivalent to the original logic.

Reply:
Note that the `return` returns from the lambda function, not `ParallelFor`.

Reply:
Thanks for clarifying, I overlooked the lambda wrapping:)